# Essential basic functionality

In [203]:
import pandas as pd
import numpy as np

In [205]:
Index = pd.date_range("05/09/2024")

ValueError: Of the four parameters: start, end, periods, and freq, exactly three must be specified

In [207]:
Index = pd.date_range("05/09/2024" , periods=8)

In [209]:
Index

DatetimeIndex(['2024-05-09', '2024-05-10', '2024-05-11', '2024-05-12',
               '2024-05-13', '2024-05-14', '2024-05-15', '2024-05-16'],
              dtype='datetime64[ns]', freq='D')

In [211]:
s = pd.Series(np.random(5), index=["a","b","c","d","e"])

TypeError: 'module' object is not callable

In [213]:
s = pd.Series(np.random.randn(5), index=["a","b","c","d","e"])

In [215]:
s

a   -0.278786
b    0.707601
c   -1.262127
d   -0.825999
e    0.502408
dtype: float64

In [217]:
df = pd.DataFrame(np.random.randn(8,3), index=Index, columns=["A","B","C"])

In [219]:
df

A         B         C
2024-05-09 -0.669373 -0.897791 -2.039007
2024-05-10  0.474056 -2.301712  0.012047
2024-05-11 -0.651026 -0.109305  0.475347
2024-05-12  1.079372  0.858243  1.103918
2024-05-13  0.367591  1.496114 -0.591148
2024-05-14  0.222514 -0.772699  0.642954
2024-05-15  0.495095 -1.283508 -0.112395
2024-05-16  2.551928  1.247187  1.468325

# Head and tail

To view a small sample of a Series or DataFrame object, use the head() and tail() methods. The default number of elements to display is five, but you may pass a custom number.

In [222]:
series = pd.Series(np.random.randn(1000))

In [224]:
series

0     -0.603932
1      0.825112
2      1.703419
3      0.358948
4      0.413435
         ...   
995    1.972014
996   -0.300611
997    1.290141
998   -0.100424
999   -0.343887
Length: 1000, dtype: float64

In [226]:
series.head()

0   -0.603932
1    0.825112
2    1.703419
3    0.358948
4    0.413435
dtype: float64

In [228]:
series.tail()

995    1.972014
996   -0.300611
997    1.290141
998   -0.100424
999   -0.343887
dtype: float64

# Attributes and underlying data

pandas object have no of attibute eanbling to acesss you metadata

shape: gives the axis dimensions of the object, consistent with ndarray

Axis labels
    Series: index (only axis)

    DataFrame: index (rows) and columns

Note, these attributes can be safely assigned to!



In [230]:
df[:2]

A         B         C
2024-05-09 -0.669373 -0.897791 -2.039007
2024-05-10  0.474056 -2.301712  0.012047

In [232]:
df.columns = [x.lower() for x in df.columns]

In [234]:
df

a         b         c
2024-05-09 -0.669373 -0.897791 -2.039007
2024-05-10  0.474056 -2.301712  0.012047
2024-05-11 -0.651026 -0.109305  0.475347
2024-05-12  1.079372  0.858243  1.103918
2024-05-13  0.367591  1.496114 -0.591148
2024-05-14  0.222514 -0.772699  0.642954
2024-05-15  0.495095 -1.283508 -0.112395
2024-05-16  2.551928  1.247187  1.468325

pandas objects (Index, Series, DataFrame) can be thought of as containers for arrays, which hold the actual da
ta and do the actual computation. For many types, the underlying array is a numpy.ndarray

To get the actual data inside a Index or Series, use the .array property

In [236]:
s.array

<NumpyExtensionArray>
[-0.2787855784414722,  0.7076012105159336, -1.2621274011981924,
 -0.8259987993942098,  0.5024081835470091]
Length: 5, dtype: float64

In [238]:
s.index.array

<NumpyExtensionArray>
['a', 'b', 'c', 'd', 'e']
Length: 5, dtype: object

array will always be an ExtensionArray. The exact details of what an ExtensionArray is and why pandas uses them are a bit beyond the scope of this introduction. See dtypes for more.

In [240]:
s.to_numpy()

array([-0.27878558,  0.70760121, -1.2621274 , -0.8259988 ,  0.50240818])

In [242]:
np.array

<function numpy.array>

In [244]:
np.asarray(s)

array([-0.27878558,  0.70760121, -1.2621274 , -0.8259988 ,  0.50240818])

When the Series or Index is backed by an ExtensionArray, to_numpy() may involve copying data and coercing values

to_numpy() gives some control over the dtype of the resulting numpy.ndarray. For example, consider datetimes with timezones. 

NumPy doesn’t have a dtype to represent timezone-aware datetimes, so there are two possibly useful representations:

An object-dtype numpy.ndarray with Timestamp objects, each with the correct tz

A datetime64[ns] -dtype numpy.ndarray, where the values have been converted to UTC and the timezone discarded

Timezones may be preserved with dtype=object



In [246]:
ser = pd.Series(pd.date_range("2024", periods=2, tz="CET" ))

In [248]:
ser

0   2024-01-01 00:00:00+01:00
1   2024-01-02 00:00:00+01:00
dtype: datetime64[ns, CET]

In [250]:
ser.to_numpy(dtype=object)

array([Timestamp('2024-01-01 00:00:00+0100', tz='CET'),
       Timestamp('2024-01-02 00:00:00+0100', tz='CET')], dtype=object)

Or thrown away with dtype='datetime64[ns]'

In [252]:
ser.to_numpy(dtype = "datetime64[ns]")

array(['2023-12-31T23:00:00.000000000', '2024-01-01T23:00:00.000000000'],
      dtype='datetime64[ns]')

Getting the “raw data” inside a DataFrame is possibly a bit more complex. When your DataFrame only has a single data type for all the columns, DataFrame.to_numpy() will return the underlying data:

In [254]:
df.to_numpy()

array([[-0.66937338, -0.89779063, -2.03900744],
       [ 0.4740555 , -2.30171153,  0.01204717],
       [-0.65102576, -0.10930545,  0.47534673],
       [ 1.07937209,  0.85824326,  1.10391811],
       [ 0.36759109,  1.49611382, -0.59114759],
       [ 0.22251376, -0.77269936,  0.64295446],
       [ 0.49509489, -1.28350795, -0.11239501],
       [ 2.55192804,  1.24718716,  1.46832454]])

If a DataFrame contains homogeneously-typed data, the ndarray can actually be modified in-place, and the changes will be reflected in the data structure. For heterogeneous data (e.g. some of the DataFrame’s columns are not all the same dtype), this will not be the case. The values attribute itself, unlike the axis labels, cannot be assigned to.

# IN the past, Pandas recommended to use Series.values and DataFrame.values to extract data from series or dataframe,
# its recommanded to not use .values insted of use a .array() .to_numpy() 

# Drowsback of .values()
    When Series contain an extention type and if used series.values() to extarct the then data or it returns Numpy array or extention array. 
    Series.array alwasy return extention array & did not copy the data
    series.to_numpy() always return numpy array & did copy/corsing the value.



    DataFrame contains a mixture of data type, DataFrame.values copy the data and corsing the value to commomn type and its reletivly it expensive operation.
    DataFrame.to_numpy() makes it cleared that return  Numpyarray 

# Accelerated operations


Pandas has support certain types of binary numerical and boolean operations using the  <b>numexpr</b>  and <b>bottleneck</b> liabraries.

<b>numexpr</b> uses smart chunking, caching, and multiple cores.

<b>bottleneck</b> is a set of specialized cython routines that are especially fast when dealing with arrays that have nans.


Here is a sample (using 100 column x 100,000 row DataFrames): df2

22.04

36.50

0.6039

b>Operation</b>  <b>0.11.0 (ms)</b> <b>Prior Version (ms)</b> <b>Ratio to Prior</b>

df1 > df           2 13.32             125.35                   0.1063

df1 * df2           21.71              36.63                    0.5928

df1 + df2          22.04               36.50                     0.6039

These are both enabled to be used by default, you can control this by setting the options:

In [256]:
bottleneckf = pd.set_option("compute.use_bottleneck", False)

In [258]:
print(bottleneckf)

None


In [260]:
bottleneckt = pd.set_option("compute.use_bottleneck", True)

In [262]:
print(bottleneckt)

None


In [264]:
numexprf = pd.set_option("compute.use_numexpr",False)

In [266]:
print(numexprf)

None


In [268]:
numexprt = pd.set_option("compute.use_numexpr",True)

In [270]:
print(numexprt)

None


# Flexible binary operations

With binary operations between pandas data structures, there are two key points of interest:

<b> Broadcasting behavior between higher- (e.g. DataFrame)</b> and <b>lower-dimensional (e.g. Series) objects</b>.

<b>Missing data in computations</b>.

We will demonstrate how to manage these issues independently, though they can be handled simultaneously.

# Matching / broadcasting behavior